### PR TITLE
Added learning rate scheduling

### PR DESCRIPTION
**Pull request by jrego43**
- Added an LRScheduler struct and Schedule trait for specifying scheduling behavior
- Implemented a variety of LR scheduler types
- Created tests for all schedulers using sample parameter values found in PyTorch examples

Source: https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate

### CODE REVIEW

File: src/lr_scheduler/cosine_annealing_lr.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            return lr;
+        }
+
+        let pi = std::f64::consts::PI;
+        let lr_range = self.base_lr - self.eta_min;
+        let wavelen_percent = pi * (last_epoch as f64) / (self.t_max as f64);
+
+        if last_epoch == 1 {
+            self.eta_min + lr_range * (1.0 + wavelen_percent.cos()) / 2.0
+        } else if (last_epoch as i32 - self.t_max as i32 - 1) % (2 * self.t_max as i32) == 0 {
```
Comment (jrego43) [inline src/lr_scheduler/cosine_annealing_lr.rs:31]: @LazaroHurtado Couldn't decide how to deal with this expression so I left it as is for this first commit.

**Comment by jrego43** [comment]
The math expressions in general are not as clear and concise as I'd like them to be, so please share suggestions regarding them.

File: examples/feedforward_nn.rs
Comment (jrego43) [inline examples/feedforward_nn.rs:1]: Meant to remove the scheduler here since it's a pretty arbitrary one originally meant for testing, but maybe it should stay or be tweaked?

File: src/lr_scheduler/mod.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    schedule: S,
+    base_lr: f64,
+    last_epoch: usize,
+}
+
+impl<S: Schedule> LRScheduler<S> {
+    pub fn new<O: Optimizer>(optimizer: &O, schedule: S) -> Self {
+        let lr = optimizer.lr();
+        let base_lr = lr.value();
+        let last_epoch = 0;
```
Comment (LazaroHurtado) [inline src/lr_scheduler/mod.rs:44]: from what I have seen, Pytorch has a default `last_epoch` value of -1, is there a reason behind using 0 here?

File: src/lr_scheduler/mod.rs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            schedule,
+            base_lr,
+            last_epoch,
+        };
+
+        scheduler.step();
+        scheduler
+    }
+
+    pub fn step(&mut self) {
```
Comment (LazaroHurtado) [inline src/lr_scheduler/mod.rs:57]: can you link Pytorch implementation of this method?

File: src/lr_scheduler/constant_lr.rs
Comment (LazaroHurtado) [inline src/lr_scheduler/constant_lr.rs:1]: Pytorch requires this schedule to be initialized with 0 <= `factor` <= 1, how is this enforced?

File: src/lr_scheduler/cosine_annealing_lr.rs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    }
+}
+
+impl Schedule for CosineAnnealingLR {
+    fn get_lr(&self, lr: f64, last_epoch: usize) -> f64 {
+        if last_epoch == 0 {
+            return lr;
+        }
+
+        let pi = std::f64::consts::PI;
```
Comment (LazaroHurtado) [inline src/lr_scheduler/cosine_annealing_lr.rs:25]: you can import this at the top of the file and just use `PI`

File: src/lr_scheduler/cosine_annealing_lr.rs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        let lr_range = self.base_lr - self.eta_min;
+        let wavelen_percent = pi * (last_epoch as f64) / (self.t_max as f64);
+
+        if last_epoch == 1 {
```
Comment (LazaroHurtado) [inline src/lr_scheduler/cosine_annealing_lr.rs:29]: Pytorch has this logic under the condition `elif self._step_count == 1 and self.last_epoch > 0:` is this accurate?

File: src/lr_scheduler/cosine_annealing_lr.rs (hunk 1 above)
Comment (LazaroHurtado) [reply src/lr_scheduler/cosine_annealing_lr.rs:31]: looks like you are type casting `t_max` twice and in other conditions, might be best to do this outside the `else if`

File: src/lr_scheduler/linear_lr.rs [hunk 6]
```diff
@@ -0,0 +1,40 @@
+use super::Schedule;
+
+pub struct LinearLR {
+    pub total_iters: usize,
+    pub start_factor: f64,
+    pub end_factor: f64,
```
Comment (LazaroHurtado) [inline src/lr_scheduler/linear_lr.rs:6]: Similiar to my other comment, there are some [constraints](https://pytorch.org/docs/stable/_modules/torch/optim/lr_scheduler.html#LinearLR) on what `start_factor` and `end_factor` can be

File: src/lr_scheduler/linear_lr.rs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+            let denominator = self.start_factor * (self.total_iters as f64)
+                + factor_range * ((last_epoch - 1) as f64);
+
+            lr * (1.0 + factor_range / denominator)
+        }
+    }
+
+    fn get_closed_form_lr(&self, base_lr: f64, last_epoch: usize) -> f64 {
+        let factor_range = self.end_factor - self.start_factor;
+        let percent_iters = (last_epoch.min(self.total_iters) as f64) / (self.total_iters as f64);
```
Comment (LazaroHurtado) [inline src/lr_scheduler/linear_lr.rs:36]: Your choice, but you can
`let total_iters = self.total_iters as f64;`
if you want instead of casting twice

File: src/lr_scheduler/multistep_lr.rs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+            Some(_) => {
+                let power =
+                    self.milestones.iter().fold(
+                        0,
+                        |acc, &epoch| if epoch == last_epoch { acc + 1 } else { acc },
+                    );
+
+                lr * self.gamma.powi(power)
+            }
+        }
```
Comment (LazaroHurtado) [inline src/lr_scheduler/multistep_lr.rs:35]:
```suggestion
        if self.milestones.contains(&last_epoch) {
                let occurrences = self.milestones.iter().filter(|&milestone| *milestone == last_epoch);
                lr * self.gamma.powi(occurrences)
        } else {
                lr        
        }
```

File: src/lr_scheduler/multistep_lr.rs [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    fn get_closed_form_lr(&self, base_lr: f64, last_epoch: usize) -> f64 {
+        let power = self.milestones.iter().fold(
+            0,
+            |acc, &epoch| if epoch <= last_epoch { acc + 1 } else { acc },
+        );
```
Comment (LazaroHurtado) [inline src/lr_scheduler/multistep_lr.rs:42]: Above comment changes can be applied here

File: src/lr_scheduler/polynomial_lr.rs [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    }
+}
+
+impl Schedule for PolynomialLR {
+    fn get_lr(&self, lr: f64, last_epoch: usize) -> f64 {
+        if (last_epoch == 0) || (last_epoch > self.total_iters) {
+            lr
+        } else {
+            let total_iters = self.total_iters as f64;
+            let base = (1.0 - (last_epoch as f64) / total_iters)
```
Comment (LazaroHurtado) [inline src/lr_scheduler/polynomial_lr.rs:23]: You can do what you did with `total_iters` to `last_epoch` here if you want

**Review by LazaroHurtado — CHANGES_REQUESTED** [review]
Big changes = Big brain, good job man! Left a few comments

File: src/lr_scheduler/constant_lr.rs
Comment (jrego43) [reply src/lr_scheduler/constant_lr.rs:1]: It’s not, will fix.

File: src/lr_scheduler/mod.rs (hunk 2 above)
Comment (jrego43) [reply src/lr_scheduler/mod.rs:44]: I made this choice early on and back then it felt like it made more sense to have `last_epoch` be a `usize` since -1 is treated as a sort of sentinel value in PyTorch. I compensated by having `step` change `lr` first and _then_ increment `last_epoch` (as opposed to PyTorch where it’s incremented first).

In retrospect, I didn’t look too far into why it’s -1 by default, and it’s likely that there’s some use case for it that I missed.

Comment (LazaroHurtado) [reply src/lr_scheduler/mod.rs:44]: Since `LRScheduler` doesn't have `last_epoch` as public it could be set to -1 but if this is a workaround thé keeping it as `usize` would be really nice

Comment (jrego43) [reply src/lr_scheduler/mod.rs:44]: I'm gonna review the PyTorch source code again and see if there's a good reason for it being -1 by default.

File: src/lr_scheduler/mod.rs (hunk 3 above)
Comment (jrego43) [reply src/lr_scheduler/mod.rs:57]: https://pytorch.org/docs/stable/_modules/torch/optim/lr_scheduler.html

Search for first instance of `def step`.

File: src/lr_scheduler/linear_lr.rs (hunk 7 above)
Comment (jrego43) [reply src/lr_scheduler/linear_lr.rs:36]: Unless I also cast `last_epoch as f64`, wouldn't that still require casting for the left side of the `percent_iters` expression?

Comment (LazaroHurtado) [reply src/lr_scheduler/linear_lr.rs:36]: Yeah good point, but now thinking about it if we change `last_epoch` to f64 to support -1 then we won't have to worry about any of these type casts for that parameter!

File: src/lr_scheduler/multistep_lr.rs (hunk 8 above)
Comment (LazaroHurtado) [reply src/lr_scheduler/multistep_lr.rs:35]: I missed the `count()`, the suggested changes should be accurate now

File: src/lr_scheduler/cosine_annealing_lr.rs (hunk 5 above)
Comment (jrego43) [reply src/lr_scheduler/cosine_annealing_lr.rs:29]: Since I did away with `step_count`, it's hard to say with absolute certainty. I would say _probably_, considering that this is in `get_lr` and not `get_closed_form_lr`. Within `get_lr`, PyTorch's `step_count` and our `last_epoch` can be treated as one and the same.

Comment (jrego43) [reply src/lr_scheduler/cosine_annealing_lr.rs:29]: We should, however, introduce `_step_count` in the next PR involving LR schedulers, as it seems to get more use when multiple schedulers for one model exist.

File: examples/feedforward_nn.rs
Comment (jrego43) [reply examples/feedforward_nn.rs:1]: @LazaroHurtado ^

File: src/value.rs [hunk 11]
```diff
@@ -103,7 +103,7 @@ impl Value {
     }
 
     pub fn powf<T: Into<f64>>(&self, raw_exponent: T) -> Self {
-        let mut exponent = Value::from(raw_exponent);
+        let mut exponent = val!(raw_exponent);
```
[the remark this answers is not in the document]
Comment (jrego43) [reply src/value.rs:106]: This would also apply to `back_pass`, right?

[the remark this answers is not in the document]
Comment (LazaroHurtado) [reply src/value.rs:106]: Yeah we could optimize at that level too! Shouldnt introduced a bug since there is no graph behind this value

File: src/optimizer/rmsprop.rs [hunk 12]
```diff
@@ -68,7 +68,7 @@ impl Optimizer for RMSProp {
 
             let momentum = self.momentum * prev_grads[i];
             prev_grads[i] = momentum + (grad / (curr_moving_avg.sqrt() + self.eps));
-            let step = self.lr * prev_grads[i];
+            let step = self.lr.value() * prev_grads[i];
```
[the remark this answers is not in the document]
Comment (jrego43) [reply src/optimizer/rmsprop.rs:71]: Going by what you say I thought removing `.value()` would be enough, but in the match arms following this line I get: 
> cannot add-assign `value::Value` to `NotNan<f64>`

File: src/lr_scheduler/constant_lr.rs
Comment (jrego43) [reply src/lr_scheduler/constant_lr.rs:1]: Since I don't have constructors for structs that implement `Schedule`, where would I enforce it? It could be done in `get_lr` and `get_closed_form_lr`, but that is obviously not ideal.

File: src/optimizer/rmsprop.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (LazaroHurtado) [reply src/optimizer/rmsprop.rs:71]: Mhm unfortunate, forgot that `value_mut` returns the f64 and not the struct itself

File: src/lr_scheduler/constant_lr.rs
Comment (LazaroHurtado) [reply src/lr_scheduler/constant_lr.rs:1]: I believe you can make the fields private and give the struct impl a `pub new` method where you will do the validation

Comment (jrego43) [reply src/lr_scheduler/constant_lr.rs:1]: Isn't `pub new` essentially a constructor method? While it would work I feel like it necessitates every `Schedule` having one so that instantiation is consistent.

Comment (LazaroHurtado) [reply src/lr_scheduler/constant_lr.rs:1]: I feel like it would be okay if only schedules that must enforce constraints have it but your choice

File: src/lr_scheduler/mod.rs (hunk 2 above)
Comment (jrego43) [reply src/lr_scheduler/mod.rs:44]: Couldn't find anything concrete, though I think we look into setting the default as -1 in a future PR that introduces more complex schedulers.

**Comment by jrego43** [comment]
It should be noted that a future scheduling-related changes is going to require reworking this PR code. For instance, some advanced schedulers require more information from the optimizer than just the learning rate.

**Review by LazaroHurtado — CHANGES_REQUESTED** [review]
LGTM but can you add some [`should_panic`](https://doc.rust-lang.org/book/ch11-01-writing-tests.html#checking-for-panics-with-should_panic) tests for LR schedulers that implement constraints on parameters?

**Review by LazaroHurtado — APPROVED** [review]
Awesome job!

**Review by Demba3 — APPROVED** [review]
LGTM
